controlplane/src/core/repositories/SchemaCheckRepository.ts

```diff
@@ -95,6 +103,109 @@ export class SchemaCheckRepository {
     await this.db.insert(schemaCheckChangeActionOperationUsage).values(values).execute();
   }
 
+  private mapChangesFromDriverValue = (val: any) => {
+    if (typeof val === 'string' && val.length > 0 && val !== '{}') {
+      const pairs = val.slice(2, -2).split('","');
+
+      return pairs.map((pair) => {
+        const [changeType, path] = pair.slice(1, -1).split(',');
+        return { changeType, path };
+      });
+    }
+    return [];
+  };
+
+  public async checkClientTrafficAgainstOverrides(data: {
+    changes: { id: string; changeType: string | null; path: string | null }[];
+    inspectorResultsByChangeId: Map<string, InspectorOperationResult[]>;
+    namespaceId: string;
+  }) {
+    let hasUnsafeClientTraffic = false;
+
+    const result = new Map(data.inspectorResultsByChangeId);
+
+    const changeActionsByOperationHash: Map<string, typeof data.changes> = new Map();
+
+    for (const [schemaCheckChangeId, operationResults] of result.entries()) {
+      for (const operationResult of operationResults) {
+        const { hash } = operationResult;
+        if (!changeActionsByOperationHash.has(hash)) {
+          changeActionsByOperationHash.set(hash, []);
+        }
+
+        const change = data.changes.find((c) => c.id === schemaCheckChangeId);
+        if (change) {
+          changeActionsByOperationHash.get(hash)?.push(change);
+        }
+      }
+    }
+
+    for (const [hash, changes] of changeActionsByOperationHash) {
+      const incomingChanges = `array[${changes.map((c) => `('${c.changeType}'::schema_change_type, '${c.path}'::text)`)}]`;
+      const storedChanges = `array_agg(distinct (${schema.operationChangeOverrides.changeType.name}, ${schema.operationChangeOverrides.path.name}))`;
+
+      // Incoming changes are new breaking changes detected in the current check run
+      // Stored changes are a a list of changes that have been marked as safe (override) by the user
+      // Here except tells us the incoming changes that are not safe and an intersect tells us the incoming changes that are safe
+      const res = await this.db
```

Comment:
This looks very complicated. We need to have at least a test that covers this.

Reply:
Done